File: backend/apiquanlynhahang/apiquanlynhahang/Controllers/BanController.cs
using apiquanlynhahang.BLL.Services;
using apiquanlynhahang.DTOs.Requests;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace apiquanlynhahang.Controllers;

[ApiController]
[Route("api/ban")]
public class BanController : ControllerBase
{
    private readonly BanService _service;

    public BanController(BanService service)
    {
        _service = service;
    }

    [AllowAnonymous]
    [HttpGet]
    public async Task<IActionResult> LayDanhSach(CancellationToken cancellationToken)
        => Ok(new { data = await _service.LayDanhSachAsync(cancellationToken) });

    [AllowAnonymous]
    [HttpGet("{maBan}")]
    public async Task<IActionResult> LayTheoMa(string maBan, CancellationToken cancellationToken)
        => Ok(new { data = await _service.LayTheoMaAsync(maBan, cancellationToken) });

    [AllowAnonymous]
    [HttpGet("{maBan}/thuc-don")]
    public async Task<IActionResult> LayThucDonTaiBan(string maBan, [FromServices] DonHangService donHangService, CancellationToken cancellationToken)
    {
        var ban = await _service.LayTheoMaAsync(maBan, cancellationToken);
        if (ban is null) return NotFound(new { message = "Bàn không tồn tại" });
        return Ok(new { ban, data = await donHangService.LayThucDonChoBanAsync(cancellationToken) });
    }

    [AllowAnonymous]
    [HttpPost("{maBan}/order")]
    public async Task<IActionResult> TaoOrderTaiBan(string maBan, [FromBody] TaoOrderTaiBanDto dto, [FromServices] DonHangService donHangService, CancellationToken cancellationToken)
    {
        var ban = await _service.LayTheoMaAsync(maBan, cancellationToken);
        if (ban is null) return NotFound(new { message = "Bàn không tồn tại" });
        return Ok(new { data = await donHangService.TaoHoacThemMonTaiBanAsync(maBan, dto, cancellationToken) });
    }

    [AllowAnonymous]
    [HttpGet("{maBan}/order")]
    public async Task<IActionResult> LayOrderDangMoTaiBan(string maBan, [FromServices] DonHangService donHangService, CancellationToken cancellationToken)
    {
        var ban = await _service.LayTheoMaAsync(maBan, cancellationToken);
        if (ban is null) return NotFound(new { message = "Bàn không tồn tại" });
        return Ok(new { data = await donHangService.LayDonTaiBanDangMoAsync(maBan, cancellationToken) });
    }

    [AllowAnonymous]
    [HttpPost("{maBan}/yeu-cau-thanh-toan")]
    public async Task<IActionResult> YeuCauThanhToanTaiBan(string maBan, [FromServices] DonHangService donHangService, CancellationToken cancellationToken)
    {
        var ban = await _service.LayTheoMaAsync(maBan, cancellationToken);
        if (ban is null) return NotFound(new { message = "Bàn không tồn tại" });
        return Ok(new { success = await donHangService.YeuCauThanhToanTaiBanAsync(maBan, cancellationToken) });
    }

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpPost("{maBan}/xac-nhan-thanh-toan")]
    public async Task<IActionResult> XacNhanThanhToanTaiBan(string maBan, [FromServices] DonHangService donHangService, CancellationToken cancellationToken)
    {
        var ban = await _service.LayTheoMaAsync(maBan, cancellationToken);
        if (ban is null) return NotFound(new { message = "Bàn không tồn tại" });
        return Ok(new { success = await donHangService.XacNhanThanhToanTaiBanAsync(maBan, cancellationToken) });
    }

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpPost]
    public async Task<IActionResult> Tao([FromBody] TaoBanDto dto, CancellationToken cancellationToken)
        => StatusCode(201, new { data = await _service.TaoAsync(dto, cancellationToken) });

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpPut("{maBan}")]
    public async Task<IActionResult> CapNhat(string maBan, [FromBody] CapNhatBanDto dto, CancellationToken cancellationToken)
    {
        var duLieu = await _service.CapNhatAsync(maBan, dto, cancellationToken);
        return duLieu is null ? NotFound(new { message = "Khong tim thay ban" }) : Ok(new { data = duLieu });
    }

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpDelete("{maBan}")]
    public async Task<IActionResult> Xoa(string maBan, CancellationToken cancellationToken)
    {
        await _service.XoaAsync(maBan, cancellationToken);
        return NoContent();
    }

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpPatch("{maBan}/status")]
    public async Task<IActionResult> CapNhatTrangThai(string maBan, [FromBody] CapNhatTrangThaiDto dto, CancellationToken cancellationToken)
        => Ok(new { data = await _service.CapNhatTrangThaiAsync(maBan, dto.TrangThai, cancellationToken) });

    [Authorize(Roles = "Admin,NhanVien")]
    [HttpGet("{maBan}/qr")]
    public async Task<IActionResult> LayQrTheoBan(string maBan, CancellationToken cancellationToken)
    {
        var scheme = Request.Scheme;
        var host = Request.Host.HasValue ? Request.Host.Value : "localhost:5173";
        var duLieu = await _service.TaoQrAsync(maBan, $"{scheme}://{host}", cancellationToken);
        return duLieu is null ? NotFound(new { message = "Khong tim thay ban" }) : Ok(new { data = duLieu });
    }
}

File: backend/nest-api/src/modules/ban/ban.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Put,
  UseGuards,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Throttle } from '@nestjs/throttler';
import { BanService } from './ban.service';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import { RolesGuard } from '../../common/guards/roles.guard';
import { Roles } from '../../common/decorators/roles.decorator';
import { Public } from '../../common/decorators/public.decorator';
import { BanGhi } from '../../common/types';

@ApiTags('ban')
@Controller('api/ban')
@UseGuards(JwtAuthGuard, RolesGuard)
export class BanController {
  constructor(private readonly banService: BanService) {}

  @Public()
  @Throttle({ default: { limit: 60, ttl: 60000 } })
  @Get()
  layDanhSachBan() {
    return this.banService.layDanhSachBan();
  }

  @Roles('Admin')
  @Post()
  taoBan(@Body() body: BanGhi) {
    return this.banService.taoBan(body);
  }

  @Roles('Admin')
  @Put(':maBan')
  capNhatBan(@Param('maBan') maBan: string, @Body() body: BanGhi) {
    return this.banService.capNhatBan(maBan, body);
  }

  @Roles('Admin')
  @Delete(':maBan')
  xoaBan(@Param('maBan') maBan: string) {
    return this.banService.xoaBan(maBan);
  }

  @Roles('Admin', 'NhanVien')
  @Patch(':maBan/status')
  capNhatTrangThaiBan(@Param('maBan') maBan: string, @Body() body: BanGhi) {
    return this.banService.capNhatTrangThaiBan(
      maBan,
      String(body.trangThai || ''),
    );
  }

  @Roles('Admin', 'NhanVien')
  @Get(':maBan/qr')
  layQrBan(@Param('maBan') maBan: string) {
    return this.banService.layQrBan(maBan);
  }

  @Public()
  @Get(':maBan/thuc-don')
  layThucDonTheoBan(@Param('maBan') maBan: string) {
    return this.banService.layThucDonTheoBan(maBan);
  }

  // Public endpoint — khách quét QR tại bàn để gọi món, không cần đăng nhập.
  // Rate limit 5 requests/phút để chống spam từ mã QR bị lộ ra ngoài.
  @Public()
  @Throttle({ default: { limit: 5, ttl: 60000 } })
  @Post(':maBan/order')
  taoOrderTaiBan(@Param('maBan') maBan: string, @Body() body: BanGhi) {
    return this.banService.taoOrderTaiBan(maBan, body);
  }

  @Public()
  @Get(':maBan/order')
  layOrderDangMoTaiBan(@Param('maBan') maBan: string) {
    return this.banService.layOrderDangMoTaiBan(maBan);
  }

  // Public endpoint — khách yêu cầu thanh toán tại bàn qua QR, không cần đăng nhập.
  // Rate limit 3 requests/phút để chống spam.
  @Public()
  @Throttle({ default: { limit: 3, ttl: 60000 } })
  @Post(':maBan/yeu-cau-thanh-toan')
  yeuCauThanhToanTaiBan(@Param('maBan') maBan: string) {
    return this.banService.yeuCauThanhToanTaiBan(maBan);
  }

  @Roles('Admin', 'NhanVien')
  @Post(':maBan/xac-nhan-thanh-toan')
  xacNhanThanhToanTaiBan(@Param('maBan') maBan: string) {
    return this.banService.xacNhanThanhToanTaiBan(maBan);
  }
}
